crates/bevy_ecs/src/system/async_system.rs

```diff
@@ -0,0 +1,586 @@
+    system_count: Arc<AtomicUsize>,
+}
+
+pub struct AsyncSystemOutput<Out: Send + Sync + 'static> {
+    rx: Receiver<Out>,
+    done: bool,
+    counter_ref: Arc<AtomicUsize>,
+}
+
+impl<In, Out> Clone for AsyncSystemHandle<In, Out> {
+    fn clone(&self) -> Self {
+        Self {
+            tx: self.tx.clone(),
+            system_count: self.system_count.clone(),
+        }
+    }
+}
+
+impl<In: Send + Sync + 'static, Out: Send + Sync + 'static> AsyncSystemHandle<In, Out> {
+    pub fn fire(&mut self, trigger: In) -> AsyncSystemOutput<Out> {
+        let (tx, rx) = async_channel::bounded(1);
+        self.tx.try_send((trigger, tx)).unwrap();
+        self.system_count.fetch_add(1, Ordering::Relaxed);
+        let counter_ref = self.system_count.clone();
+        AsyncSystemOutput {
+            rx,
+            done: false,
+            counter_ref,
+        }
+    }
+
+    pub fn active_system_count(&self) -> usize {
+        self.system_count.load(Ordering::Relaxed)
+    }
+}
+
+impl<Out: Send + Sync + 'static> AsyncSystemOutput<Out> {
+    pub fn get(&mut self) -> Result<Out, AsyncSystemOutputError> {
+        match self.rx.try_recv() {
+            Ok(v) => {
+                self.counter_ref.fetch_sub(1, Ordering::Relaxed);
+                Ok(v)
+            }
+            Err(async_channel::TryRecvError::Empty) => Err(if self.done {
+                AsyncSystemOutputError::OutputMoved
+            } else {
+                AsyncSystemOutputError::SystemNotFinished
+            }),
+            Err(async_channel::TryRecvError::Closed) => panic!(),
+        }
+    }
+
+    pub fn check(&self) -> bool {
+        !self.rx.is_empty()
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum AsyncSystemOutputError {
+    #[error("The output of this system call has already been taken")]
+    OutputMoved,
+    #[error("The system has not finished")]
+    SystemNotFinished,
+}
+
+pub struct Accessor<P: SystemParam> {
+    channel: Sender<Box<dyn FnOnce(&SystemState, &World, &Resources) + Send + Sync>>,
+    _marker: OpaquePhantomData<P>,
+}
+
+impl<P: SystemParam> Clone for Accessor<P> {
+    fn clone(&self) -> Self {
+        Self {
+            channel: self.channel.clone(),
+            _marker: Default::default(),
+        }
+    }
+}
+
+impl<P: SystemParam> Accessor<P> {
+    pub async fn access<F, R: Send + 'static>(&mut self, sync: F) -> R
+    where
+        // Removing the 'static here would allow removing the
+        // transmutes, but its currently not possible due to an ICE.
+        F: FnOnce(<P::Fetch as FetchSystemParam<'static>>::Item) -> R + Send + Sync + 'static,
+    {
+        let (tx, rx) = async_channel::bounded(1);
+        self.channel
+            .send(Box::new(move |state, world, resources| {
+                // Safe: the sent closure is executed inside run_unsafe, which provides the correct guarantees.
+                if let Some(params) = unsafe {
+                    P::Fetch::get_param(
```

Comment:
What if `P::Fetch::get_param` stores any of these in a static?

Reply:
I don't fully know why, but it thankfully doesn't work.
```rust
lazy_static::lazy_static! {
    static ref TEST: Mutex<Option<&'static mut str>> = Default::default();
}

async fn compile_fail(mut access: Accessor<ResMut<'_, String>>) {
    access
        .access(|mut res| {
            let mutable = &mut *res;
            let test = &mut *TEST.lock();
            test.replace(mutable);
        })
        .await;
}
```
This snippet fails to compile with the message `res does not live long enough`

Reply:
Hmm, that would be unsound behavior. I hadn't been able to make that happen yet, but now I did and I'm sad.
```rust
thread_local! {
    static TEST: Mutex<Option<ResMut<'static, String>>> = Default::default();
}
async fn compile_fail(mut access: Accessor<ResMut<'_, String>>) {
    access
        .access(|res| {
            TEST.with(|mutex| {
                let test = &mut *mutex.lock();
                test.replace(res);
            });
        })
        .await;
}
```
This code unfortunately compiles.

Reply:
```rust
pub async fn access<F, R: Send + 'static>(&mut self, sync: F) -> R
where
    F: FnOnce(<P::Fetch as FetchSystemParam>::Item) -> R + Send + Sync + 'static,
{
    let (tx, rx) = async_channel::bounded(1);
    self.channel
        .send(Box::new(move |state, world, resources| {
            // Safe: the sent closure is executed inside run_unsafe, which provides the correct guarantees.
            if let Some(params) = unsafe { P::Fetch::get_param(state, world, resources) } {
                tx.try_send(sync(params)).unwrap();
            }
        }))
        .await
        .unwrap();
    rx.recv().await.unwrap()
}
```
This version of the code avoids this unsound behavior, but unfortunately it fails when compiling due to an ICE